flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBKeyedStateBackend.java

```diff
@@ -718,70 +718,63 @@ void restore(Collection<KeyedStateHandle> restoreStateHandles) throws Exception
 			boolean hasExtraKeys = (restoreStateHandles.size() > 1 ||
 				!Objects.equals(restoreStateHandles.iterator().next().getKeyGroupRange(), stateBackend.keyGroupRange));
 
-			if (hasExtraKeys) {
-				stateBackend.createDB();
-			}
-
-			for (KeyedStateHandle rawStateHandle : restoreStateHandles) {
-
-				if (rawStateHandle instanceof IncrementalKeyedStateHandle) {
-					restoreInstance((IncrementalKeyedStateHandle) rawStateHandle, hasExtraKeys);
-				} else if (rawStateHandle instanceof IncrementalLocalKeyedStateHandle) {
-					Preconditions.checkState(!hasExtraKeys, "Cannot recover from local state after rescaling.");
-					restoreInstance((IncrementalLocalKeyedStateHandle) rawStateHandle);
-				} else {
-					throw new IllegalStateException("Unexpected state handle type, " +
-						"expected " + IncrementalKeyedStateHandle.class +
-						", but found " + rawStateHandle.getClass());
-				}
+			if (!hasExtraKeys) {
+				restoreFromSingleHandle(restoreStateHandles.iterator().next());
```

Comment:
This new (and also the old code before) look like there could be a potential bug: if `restoreStateHandles.size() > 1` is false, how can we be sure that `restoreStateHandles.iterator().next()` exists? Even if it works from some hidden assumption, it does not look so clean.

Reply:
You are right, I think this should be improved.